File: src/demo2_apriltag_docking/test/test_quality_policy.py
import math

from demo2_apriltag_docking.quality_policy import (
    pose_delta,
    validate_camera_calibration,
    validate_observation_timing,
)
import pytest


def test_accepts_fresh_detection_and_transform():
    result = validate_observation_timing(
        now=10.0,
        detection_stamp=9.9,
        tf_stamp=9.85,
        max_detection_age=0.25,
        max_tf_age=0.25,
        max_future_skew=0.05,
    )

    assert result.accepted is True
    assert result.detection_age_ms == pytest.approx(100.0)
    assert result.tf_age_ms == pytest.approx(150.0)


def test_rejects_stale_and_future_temporal_inputs():
    stale = validate_observation_timing(
        now=10.0,
        detection_stamp=9.7,
        tf_stamp=9.9,
        max_detection_age=0.25,
        max_tf_age=0.25,
        max_future_skew=0.05,
    )
    future = validate_observation_timing(
        now=10.0,
        detection_stamp=10.1,
        tf_stamp=10.0,
        max_detection_age=0.25,
        max_tf_age=0.25,
        max_future_skew=0.05,
    )

    assert stale.reason == 'DETECTION_STALE'
    assert future.reason == 'DETECTION_FUTURE'


def test_validates_metric_camera_intrinsics():
    valid = validate_camera_calibration(
        640,
        480,
        (500.0, 0.0, 320.0, 0.0, 500.0, 240.0, 0.0, 0.0, 1.0),
        'plumb_bob',
    )
    invalid = validate_camera_calibration(
        640,
        480,
        (0.0, 0.0, 320.0, 0.0, 0.0, 240.0, 0.0, 0.0, 1.0),
        'plumb_bob',
    )

    assert valid.valid is True
    assert valid.fx == 500.0
    assert invalid.reason == 'FOCAL_LENGTH_INVALID'


def test_rejects_missing_intrinsics_without_raising():
    result = validate_camera_calibration(640, 480, None, 'plumb_bob')

    assert result.valid is False
    assert result.reason == 'INTRINSICS_INVALID'


def test_pose_delta_wraps_yaw_across_pi():
    translation, yaw = pose_delta(
        (0.0, 0.0, 0.0, math.radians(179.0)),
        (0.1, 0.0, 0.0, math.radians(-179.0)),
    )

    assert translation == 0.1
    assert math.isclose(yaw, math.radians(2.0))

File: src/demo2_apriltag_docking/test/test_quality_policy.cpp
#include "demo2_apriltag_docking/quality_policy.hpp"

#include <gtest/gtest.h>

#include <cmath>

namespace demo2_apriltag_docking
{
namespace
{

TEST(QualityPolicy, AcceptsFreshObservation)
{
  const auto result = validate_observation_timing(10.0, 9.9, 9.85, 0.25, 0.25, 0.05);
  EXPECT_TRUE(result.accepted);
  EXPECT_NEAR(*result.detection_age_ms, 100.0, 1e-9);
  EXPECT_NEAR(*result.tf_age_ms, 150.0, 1e-9);
}

TEST(QualityPolicy, RejectsStaleAndFutureObservation)
{
  EXPECT_STREQ(
    validate_observation_timing(10.0, 9.7, 9.9, 0.25, 0.25, 0.05).reason,
    "DETECTION_STALE");
  EXPECT_STREQ(
    validate_observation_timing(10.0, 10.1, 10.0, 0.25, 0.25, 0.05).reason,
    "DETECTION_FUTURE");
}

TEST(QualityPolicy, ValidatesCameraCalibration)
{
  const std::array<double, 9> matrix{500.0, 0.0, 320.0, 0.0, 500.0, 240.0, 0.0, 0.0, 1.0};
  const auto valid = validate_camera_calibration(640, 480, matrix, "plumb_bob");
  EXPECT_TRUE(valid.valid);
  EXPECT_DOUBLE_EQ(*valid.fx, 500.0);

  const std::array<double, 9> invalid_matrix{0.0, 0.0, 320.0, 0.0, 0.0, 240.0, 0.0, 0.0, 1.0};
  EXPECT_STREQ(
    validate_camera_calibration(640, 480, invalid_matrix, "plumb_bob").reason,
    "FOCAL_LENGTH_INVALID");
}

TEST(QualityPolicy, WrapsYawDeltaAcrossPi)
{
  constexpr double pi = 3.14159265358979323846;
  const auto [translation, yaw] = pose_delta(
    std::array<double, 4>{0.0, 0.0, 0.0, 179.0 * pi / 180.0},
    std::array<double, 4>{0.1, 0.0, 0.0, -179.0 * pi / 180.0});
  EXPECT_DOUBLE_EQ(translation, 0.1);
  EXPECT_NEAR(yaw, 2.0 * pi / 180.0, 1e-12);
}

}  // namespace
}  // namespace demo2_apriltag_docking
